sphinxgallery/gen_rst.py

```diff
@@ -292,157 +254,175 @@ def scale_image(in_fname, out_fname, max_width, max_height):
+
+    except Exception:
+        image_list = '%s is not compiling:' % src_file
+        print(80 * '_')
+        print(image_list)
+        traceback.print_exc()
+        print(80 * '_')
+    finally:
+        os.chdir(cwd)
+        sys.stdout = orig_stdout
+
+    print(" - time elapsed : %.2g sec" % time_elapsed)
+    code_output = "\n{0}\n\n{1}\n\n".format(image_list, stdout)
 
+    return code_output, time_elapsed, fig_count + len(figure_list)
 
-def generate_file_rst(fname, target_dir, src_dir, plot_gallery):
+
+def generate_file_rst(fname, target_dir, src_dir):
     """ Generate the rst file for a given example."""
-    base_image_name = os.path.splitext(fname)[0]
-    image_fname = 'sphx_glr_%s_%%03d.png' % base_image_name
 
-    this_template = rst_template
-    short_fname = target_dir.replace(os.path.sep, '_') + '_' + fname
     src_file = os.path.join(src_dir, fname)
     example_file = os.path.join(target_dir, fname)
     shutil.copyfile(src_file, example_file)
 
     image_dir = os.path.join(target_dir, 'images')
-    thumb_dir = os.path.join(image_dir, 'thumb')
-    thumb_file = os.path.join(thumb_dir, 'sphx_glr_%s_thumb.png' % base_image_name)
     if not os.path.exists(image_dir):
         os.makedirs(image_dir)
-    if not os.path.exists(thumb_dir):
-        os.makedirs(thumb_dir)
+
+    base_image_name = os.path.splitext(fname)[0]
+    image_fname = 'sphx_glr_' + base_image_name + '_{0:03}.png'
+    image_path = os.path.join(image_dir, image_fname)
+
+    if _plots_are_current(src_file, image_path):
+        return
 
     time_elapsed = 0
-    if plot_gallery and fname.startswith('plot'):
-        # generate the plot as png image if file name
-        # starts with plot and if it is more recent than an
-        # existing image.
-        image_list, time_elapsed, stdout = execute_script(image_dir,
-                                                          thumb_file,
-                                                          image_fname,
-                                                          base_image_name,
-                                                          src_file, fname)
-        this_template = plot_rst_template
-
-    if not os.path.exists(thumb_file):
-        # create something to replace the thumbnail
-        scale_image(os.path.join(glr_path_static(), 'no_image.png'),
-                    thumb_file, 200, 140)
+    script_blocks = split_code_and_text_blocks(example_file)
+
+    ref_fname = example_file.replace(os.path.sep, '_')
+    example_rst = """\n\n.. _sphx_glr_{0}:\n\n""".format(ref_fname)
+
+    if not fname.startswith('plot'):
+        convert_func = dict(code=codestr2rst, text=text2string)
+        for blabel, bcontent in script_blocks:
+            example_rst += convert_func[blabel](bcontent)+'\n'
+    else:
+        example_globals = {}
+        fig_count = 0
+        for blabel, bcontent in script_blocks:
+            if blabel == 'code':
```

Comment:
slight niggle: you can take that out of the loop:

```python
code_blocks = [label, content for label, content in script_blocks if label == 'code']
for blabel, bcontent in code_blocks:
    ...
```

Now you can even use len(code_blocks) > 1 as the is_example_notebook_like condition which seems a tad less magical.